pkg/provisioner/crdprovisioner.go

```diff
@@ -375,72 +380,60 @@ func (c *CrdProvisioner) PublishVolume(
 	}
 	volumeName = strings.ToLower(volumeName)
 
-	attachmentName := azureutils.GetAzVolumeAttachmentName(volumeName, nodeID)
-
-	waiter, err := c.conditionWatcher.newConditionWaiter(ctx, azVolumeAttachmentType, attachmentName, func(obj interface{}, _ bool) (bool, error) {
-		if obj == nil {
-			return false, nil
-		}
-		azVolumeAttachmentInstance := obj.(*diskv1alpha2.AzVolumeAttachment)
-		if azVolumeAttachmentInstance.Status.Detail != nil && azVolumeAttachmentInstance.Status.Detail.PublishContext != nil && azVolumeAttachmentInstance.Status.Detail.Role == diskv1alpha2.PrimaryRole {
-			return true, nil
-		}
-		if azVolumeAttachmentInstance.Status.Error != nil {
-			return false, util.ErrorFromAzError(azVolumeAttachmentInstance.Status.Error)
-		}
-		return false, nil
-	})
+	// return error if volume is not found
+	azVolume, err := volumeLister.Get(volumeName)
+	if errors.IsNotFound(err) {
+		return nil, status.Errorf(codes.NotFound, "volume (%s) does not exist", volumeName)
+	}
 
-	if err != nil {
-		return nil, err
+	// return error if node is not found
+	if _, err := nodeLister.Get(nodeID); errors.IsNotFound(err) {
+		return nil, status.Errorf(codes.NotFound, "node (%s) does not exist", nodeID)
 	}
-	defer waiter.Close()
 
 	if volumeContext == nil {
 		volumeContext = map[string]string{}
 	}
-	azVolumeAttachmentInstance, err := lister.Get(attachmentName)
-	if err == nil {
-		// if CRI is scheduled for deletion, retry until operations are complete
-		if azVolumeAttachmentInstance.DeletionTimestamp != nil {
-			return nil, status.Errorf(codes.Aborted, "need to wait until attachment is fully deleted for volume (%s) and node (%s) before attaching", volumeName, nodeID)
+
+	publishContext := map[string]string{}
+
+	// preemptively detach any replica volume if necessary
+	volumeLabel, _ := azureutils.CreateLabelRequirements(consts.VolumeNameLabel, selection.Equals, volumeName)
+	nodeLabel, _ := azureutils.CreateLabelRequirements(consts.NodeNameLabel, selection.NotEquals, nodeID)
+	azVASelector := labels.NewSelector().Add(*volumeLabel, *nodeLabel)
+	existingAttachments, err := azVALister.List(azVASelector)
```

Comment:
I believe you are assuming that all other attachments for this volume not on this node are replicas. However, this assumption doesn't hold for RWX and RWO block volumes with shared mounts. In this case, PublishVolume will be called to create multiple primary attachments. We'll need to further filter this list to exclude other primary mounts.

Reply:
Hmm. I guess this is handled by the check for `azVolume.Spec.MaxMountReplicaCount > 0` below, but it would be safer to exclude primary mounts in this list anyway.

Reply:
Maybe a better and more efficient way to handle this is to only enumerate the other `AzVolumeAttachments` if `azVolume.Spec.MaxMountReplicaCount > 0` (i.e. enclose lines 400 through 424 in `if azVolume.Spec.MaxMountReplicaCount > 0 {...}`).